servant-server/src/Servant/Server/Internal/Router.hs

```diff
@@ -55,18 +59,23 @@ runRouter (StaticRouter table) request respond =
       | Just router <- M.lookup first table
       -> let request' = request { pathInfo = rest }
          in  runRouter router request' respond
-    _ -> respond $ failWith NotFound
+    _ -> respond $ Fail err404
 runRouter (DynamicRouter fun)  request respond =
   case processedPathInfo request of
     first : rest
       -> let request' = request { pathInfo = rest }
          in  runRouter (fun first) request' respond
-    _ -> respond $ failWith NotFound
+    _ -> respond $ Fail err404
 runRouter (LeafRouter app)     request respond = app request respond
 runRouter (Choice r1 r2)       request respond =
-  runRouter r1 request $ \ mResponse1 ->
-    if isMismatch mResponse1
-      then runRouter r2 request $ \ mResponse2 ->
-             respond (mResponse1 <> mResponse2)
-      else respond mResponse1
-
+  runRouter r1 request $ \ mResponse1 -> case mResponse1 of
+    Fail _ -> runRouter r2 request $ \ mResponse2 ->
+      respond (highestPri mResponse1 mResponse2)
+    _      -> respond mResponse1
+  where
+    highestPri (Fail e1) (Fail e2) =
+      if errHTTPCode e1 == 404 && errHTTPCode e2 /= 404
+        then Fail e2
+        else Fail e1
+    highestPri (Fail _) y = y
+    highestPri x _ = x
```

Comment:
This is the point where `FailFatal` is actually treated as being fatal, right? This should be documented, and possibly be made explicit in pattern matching rather than hidden in a catch-all.